apps/rush-lib/src/api/CommandLineConfiguration.ts

```diff
@@ -29,270 +96,399 @@ export class CommandLineConfiguration {
+              i--;
+              populateCommandAssociatedParametersForPhase(syntheticPhaseName, normalizedParameter);
+              parameterHasAssociations = true;
+            } else if (!this.commands.has(associatedCommandName)) {
               throw new Error(
-                `${RushConstants.commandLineFilename} defines a parameter "${parameter.longName}" ` +
-                  `that is associated with a phase "${associatedPhase}" that does not exist.`
+                `${RushConstants.commandLineFilename} defines a parameter "${normalizedParameter.longName}" ` +
+                  `that is associated with a command "${associatedCommandName}" that does not exist or does ` +
+                  'not support custom parameters.'
               );
             } else {
+              const associatedCommand: Command = this.commands.get(associatedCommandName)!;
+              associatedCommand.associatedParameters.add(normalizedParameter);
               parameterHasAssociations = true;
             }
           }
+        }
 
-          if (!parameterHasAssociations) {
+        for (const associatedPhase of normalizedParameter.associatedPhases || []) {
+          if (!this.phases.has(associatedPhase)) {
             throw new Error(
-              `${RushConstants.commandLineFilename} defines a parameter "${parameter.longName}"` +
-                ` that lists no associated commands or phases.`
+              `${RushConstants.commandLineFilename} defines a parameter "${normalizedParameter.longName}" ` +
+                `that is associated with a phase "${associatedPhase}" that does not exist.`
             );
+          } else {
+            populateCommandAssociatedParametersForPhase(associatedPhase, normalizedParameter);
+            parameterHasAssociations = true;
           }
         }
+
+        if (!parameterHasAssociations) {
+          throw new Error(
+            `${RushConstants.commandLineFilename} defines a parameter "${normalizedParameter.longName}"` +
+              ` that lists no associated commands or phases.`
+          );
+        }
       }
     }
   }
 
-  private _checkForPhaseSelfCycles(phase: IPhaseJson, checkedPhases: Set<string> = new Set<string>()): void {
+  private _checkForPhaseSelfCycles(phase: IPhase, checkedPhases: Set<string> = new Set<string>()): void {
     const dependencies: string[] | undefined = phase.dependencies?.self;
     if (dependencies) {
       for (const dependencyName of dependencies) {
         if (checkedPhases.has(dependencyName)) {
+          const dependencyNameForError: string =
+            typeof dependencyName === 'string' ? dependencyName : '<synthetic phase>';
           throw new Error(
             `In ${RushConstants.commandLineFilename}, there exists a cycle within the ` +
-              `set of ${dependencyName} dependencies: ${Array.from(checkedPhases).join(', ')}`
+              `set of ${dependencyNameForError} dependencies: ${Array.from(checkedPhases).join(', ')}`
```

Comment:
This message seems to be printing a list of dependencies (i.e. Rush project names?), however the variable is called `checkedPhases`?  Is the code correct?  If so, could the variable name be more clear?

Reply:
"dependencies" here refers to dependencies within phases. Renamed `dependencies` to `phaseSelfDependencies`